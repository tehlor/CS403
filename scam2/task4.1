(define sd '(define (f x) (let* ((y 2)) (+ x y))))
(println (let*->lambdas sd))
(inspect ((eval (let*->lambdas sd) this) 3))
