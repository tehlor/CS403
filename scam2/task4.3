(println (let*->lambdas '(define (g x y) (let* ((w (+ 2 3))) (+ x y w)))))
