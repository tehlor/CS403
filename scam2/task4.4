(println (let*->lambdas '(define (g x y) (let* ((w (+ 2 3))) (inspect w) (+ x y w)))))
