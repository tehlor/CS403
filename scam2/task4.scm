(define (main)
    (setPort (open (getElement ScamArgs 1) 'read))
    (readArgs)
)

(define (readArgs)
    (define env this)
    (define (iter expr)
        (if (not (eof?))
            (begin (eval expr env) 
                (iter (readExpr))
            )
        )
    )
    (iter (readExpr))
)

(define (let*->lambdas x)
    (if (eq? 'let* (car (car (cdr (cdr x)))))
        (list
        'define
        (car (cdr x))
        ;(car (cdr (cdr x))) Let expression
        ;(car (car (cdr (cdr x)))) ;let*
        (construct (cdr (car (cdr (cdr x)))))
        ;(cdr (car (cdr (cdr x)))) ; without let ;list of variables and values
        ;(car (cdr (cdr (car (cdr (cdr x)))))) ;list of functions
        )
        x
    )
)

(define (construct x)
    ;(println x)
    (if (< (length (car x)) 2)
                    (append (list
                    (append (list 
                    'lambda 
                    (variables (car x)))
                    (cdr x)))
                    (values (car x)))
            (append (list 
                    (list
                    'lambda 
                    (variables (car x))
                    (construct (append (list(cdr (car x))) (cdr x)))))
                    (values (car x)))))


(define (variables l)
    ;(println "Length : " (length l) "; Variables: "l)
    (if (< (length l) 2)
        (list (car (car l)))
        (variables (list(car l)))
    )
)

(define (values l)
    ;(println "Length : " (length l) "; Values: "l)
    (if (< (length l) 2)
        (list (car (cdr (car l))))
        (variables (list(cdr (car l))))
    )
)