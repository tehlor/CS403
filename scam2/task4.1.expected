(define (f x) ((lambda (y) (+ x y)) 2))
((eval (let*->lambdas sd) this) 3) is 5
